vault/identity_store_oidc_provider.go

```diff
@@ -0,0 +1,164 @@
+package vault
+
+import (
+	"context"
+
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+type assignment struct {
+	name     string
+	Groups   []string `json:"groups"`
+	Entities []string `json:"entities"`
+}
+
+const (
+	oidcProviderPrefix = "oidc_provider/"
+	assignmentPath     = oidcProviderPrefix + "assignment/"
+)
+
+func oidcProviderPaths(i *IdentityStore) []*framework.Path {
+	return []*framework.Path{
+		{
+			Pattern: "oidc/assignment/" + framework.GenericNameRegex("name"),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the assignment",
+				},
+				"entities": {
+					Type:        framework.TypeCommaStringSlice,
+					Description: "Comma separated string or array of identity entity names",
+				},
+				"groups": {
+					Type:        framework.TypeCommaStringSlice,
+					Description: "Comma separated string or array of identity group names",
+				},
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.UpdateOperation: &framework.PathOperation{
+					Callback: i.pathOIDCCreateUpdateAssignment,
+				},
+				logical.CreateOperation: &framework.PathOperation{
+					Callback: i.pathOIDCCreateUpdateAssignment,
+				},
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: i.pathOIDCReadAssignment,
+				},
+				logical.DeleteOperation: &framework.PathOperation{
+					Callback: i.pathOIDCDeleteAssignment,
+				},
+			},
+			ExistenceCheck:  i.pathOIDCAssignmentExistenceCheck,
+			HelpSynopsis:    "CRUD operations for OIDC assignments.",
+			HelpDescription: "Create, Read, Update, and Delete OIDC assignments.",
+		},
+		{
+			Pattern: "oidc/assignment/?$",
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ListOperation: &framework.PathOperation{
+					Callback: i.pathOIDCListAssignment,
+				},
+			},
+			HelpSynopsis:    "List OIDC assignments",
+			HelpDescription: "List all configured OIDC assignments in the identity backend.",
+		},
+	}
+}
+
+// pathOIDCCreateUpdateAssignment is used to create a new assignment or update an existing one
+func (i *IdentityStore) pathOIDCCreateUpdateAssignment(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	var assignment assignment
+	if req.Operation == logical.UpdateOperation {
+		entry, err := req.Storage.Get(ctx, assignmentPath+name)
+		if err != nil {
+			return nil, err
+		}
+		if entry != nil {
+			if err := entry.DecodeJSON(&assignment); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if entitiesRaw, ok := d.GetOk("entities"); ok {
+		assignment.Entities = entitiesRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		assignment.Entities = d.GetDefaultOrZero("entities").([]string)
```

Comment:
There doesn't seem to be a default FieldSchema value for `entries` and `groups`. We can just do `d.Get` in these instances.

Reply:
The reason we went with `GetDefaultOrZero` was to be a bit more explicit about what is happening in this block (getting the zero value). Here is some context: https://github.com/hashicorp/vault/pull/12198#discussion_r687783590

But `d.Get` is the common pattern used in this scenario.

Reply:
I'm a bit confused on how `GetDefaultOrZero` can [return the actual value provided](https://github.com/hashicorp/vault/blob/861454e0ed1390d67ddaf1a53c1798e5e291728c/sdk/framework/field_data.go#L79). IIUC it always either returns the default (if one is set in the schema) or the zero value. We'd have to use `d.Get` in order to properly parse user-provided values, but then how come the tests work fine?

Reply:
Oh I see why! Try adding a `logical.CreateOperation` test case with non-empty `"groups"` or `"entities"` and check by reading it right after. It should fail because the group will always be empty due to `GetDefaultOrZero`.

Reply:
Nvm, the if case covers this. The overall logic is fine, just reads a bit odd.

Reply:
We noticed that the call to `d.Get()` in the `else if` branch never returns the actual value provided. It always hits [GetDefaultOrZero](https://github.com/hashicorp/vault/blob/main/sdk/framework/field_data.go#L70). If the value was provided, it would've already been set by the conditional `d.GetOk()` branch above. See that `d.Get()` just calls [d.GetOk()](https://github.com/hashicorp/vault/blob/main/sdk/framework/field_data.go#L68).